iceoryx_posh/source/runtime/service_discovery.cpp

```diff
@@ -23,40 +23,54 @@ namespace runtime
 {
 ServiceContainer ServiceDiscovery::findService(const cxx::optional<capro::IdString_t>& service,
                                                const cxx::optional<capro::IdString_t>& instance,
-                                               const cxx::optional<capro::IdString_t>& event) noexcept
+                                               const cxx::optional<capro::IdString_t>& event,
+                                               const MessagingPattern pattern) noexcept
 {
-    m_serviceRegistrySubscriber.take().and_then([&](popo::Sample<const roudi::ServiceRegistry>& serviceRegistrySample) {
-        m_serviceRegistry = *serviceRegistrySample;
-    });
-
-    roudi::ServiceRegistry::ServiceDescriptionVector_t tempSearchResult;
-    m_serviceRegistry.find(tempSearchResult, service, instance, event);
-
     ServiceContainer searchResult;
-    for (auto& service : tempSearchResult)
-    {
-        if (service.publisherCount > 0)
-        {
-            searchResult.push_back(service.serviceDescription);
-        }
-    }
+
+    auto lambda = [&](const capro::ServiceDescription& entry) { searchResult.emplace_back(entry); };
+    findService(service, instance, event, lambda, pattern);
 
     return searchResult;
 }
 
-void ServiceDiscovery::findService(
-    const cxx::optional<capro::IdString_t>& service,
-    const cxx::optional<capro::IdString_t>& instance,
-    const cxx::optional<capro::IdString_t>& event,
-    const cxx::function_ref<void(const capro::ServiceDescription&)>& callableForEach) noexcept
+void ServiceDiscovery::findService(const cxx::optional<capro::IdString_t>& service,
+                                   const cxx::optional<capro::IdString_t>& instance,
+                                   const cxx::optional<capro::IdString_t>& event,
+                                   const cxx::function_ref<void(const capro::ServiceDescription&)>& callableForEach,
+                                   const MessagingPattern pattern) noexcept
 {
     if (!callableForEach)
     {
         return;
     }
 
-    auto lambda = [&](const roudi::ServiceRegistry::ServiceDescriptionEntry& s) {
-        callableForEach(s.serviceDescription);
+    auto lambda = [&](const roudi::ServiceRegistry::ServiceDescriptionEntry& serviceEntry) {
```

Comment:
Please create the lambda based on the case, do not resolve the case in the lambda.

Reply:
That's more nifty, done.